### PR TITLE
feat: add `Texture2DArray` feat

### PR DESCRIPTION
**Pull request by GuoLei1990**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
Add Texture2DArray.

### CODE REVIEW

**Review by singlecoder — APPROVED** [review]
+1

File: packages/rhi-webgl/src/GLTexture2DArray.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    if (formatDetail.isCompressed) {
+      throw new Error("Unable to read compressed texture");
+    }
+
+    gl.bindFramebuffer(gl.FRAMEBUFFER, this._getReadFrameBuffer());
+    gl.framebufferTextureLayer(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, this._glTexture, mipLevel, elementIndex);
+    gl.readPixels(x, y, width, height, formatDetail.baseFormat, formatDetail.dataType, out);
+    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
+  }
```
Comment (zhuxudong) [inline packages/rhi-webgl/src/GLTexture2DArray.ts:110]: why not refactor `GLTexture._getPixelBuffer` ?

File: packages/core/src/renderingHardwareInterface/IPlatformTexture2DArray.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import { IPlatformTexture } from "./IPlatformTexture";
+
+/**
+ * 2D texture array interface specification.
+ */
+export interface IPlatformTexture2DArray extends IPlatformTexture {
+  /**
+   * Setting pixels data through color buffer data, designated area and texture mipmapping level,it's also applicable to compressed formats.
+   * @remarks If it is the WebGL1.0 platform and the texture format is compressed, the first upload must be filled with textures.
```
Comment (zhuxudong) [inline packages/core/src/renderingHardwareInterface/IPlatformTexture2DArray.ts:9]: delete it cause is always 2.0

**Review by zhuxudong — CHANGES_REQUESTED** [review]

File: packages/rhi-webgl/src/GLTexture2DArray.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    height = height || Math.max(1, this._texture.height >> mipLevel) - y;
+
+    this._bind();
+    gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, 0);
+    gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, 0);
+
+    if (isCompressed) {
+      gl.compressedTexSubImage3D(target, mipLevel, x, y, elementIndex, width, height, 1, internalFormat, colorBuffer);
+    } else {
+      gl.texSubImage3D(target, mipLevel, x, y, elementIndex, width, height, 1, baseFormat, dataType, colorBuffer);
```
Comment (zhuxudong) [inline packages/rhi-webgl/src/GLTexture2DArray.ts:51]: should be `texSubImage3D(target, mipLevel, x, y, 0, width, height, elementIndex, baseFormat, dataType, colorBuffer);`

File: packages/rhi-webgl/src/GLTexture2DArray.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+    width = width || Math.max(1, this._texture.width >> mipLevel) - x;
+    height = height || Math.max(1, this._texture.height >> mipLevel) - y;
+
+    this._bind();
+    gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, 0);
+    gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, 0);
+
+    if (isCompressed) {
+      gl.compressedTexSubImage3D(target, mipLevel, x, y, elementIndex, width, height, 1, internalFormat, colorBuffer);
```
Comment (zhuxudong) [inline packages/rhi-webgl/src/GLTexture2DArray.ts:49]: should be `gl.compressedTexSubImage3D(target, mipLevel, x, y, 0, width, height, elementIndex, internalFormat, colorBuffer);`

File: packages/rhi-webgl/src/GLTexture2DArray.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+      x,
+      y,
+      elementIndex,
+      imageSource.width,
+      imageSource.height,
+      1,
+      baseFormat,
+      dataType,
+      imageSource
+    );
```
Comment (zhuxudong) [inline packages/rhi-webgl/src/GLTexture2DArray.ts:85]: should be ` gl.texSubImage3D(
      this._target,
      mipLevel,
      x,
      y,
      0,
      imageSource.width,
      imageSource.height,
      elementIndex,
      baseFormat,
      dataType,
      imageSource
    );`

**Review by zhuxudong — CHANGES_REQUESTED** [review]

**Comment by GuoLei1990** [comment]
https://github.com/oasis-engine/engine/pull/737 inclue this.

File: packages/rhi-webgl/src/GLTexture2DArray.ts (hunk 1 above)
Comment (GuoLei1990) [reply packages/rhi-webgl/src/GLTexture2DArray.ts:110]: Will cause _getPixelBuffer to be too complicated, `Texture2DArray` uses framebufferTextureLayer, there is no reuse code

**Pull request closed without merging**